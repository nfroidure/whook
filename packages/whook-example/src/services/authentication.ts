import { autoService, location } from 'knifecycle';
import { YError } from 'yerror';
import { type JWTService } from 'jwt-service';
import {
  type AuthenticationService,
  type BaseAuthenticationData,
} from '@whook/authorization';

export default location(autoService(initAuthentication), import.meta.url);

export type AuthenticationDependencies = {
  jwtToken: JWTService<AuthenticationData>;
};

export type AuthenticationData = BaseAuthenticationData & {
  userId: string;
};

type FakePayload = AuthenticationData;
type BearerPayload = { hash: string };

/* Architecture Note #4.1: authentication

A fake authentication service you can use as a base
 authentication service.
*/
async function initAuthentication({
  jwtToken,
}: AuthenticationDependencies): Promise<
  AuthenticationService<FakePayload | BearerPayload, AuthenticationData>
> {
  const authentication = {
    check: async (
      type: string,
      data: FakePayload | BearerPayload,
    ): Promise<AuthenticationData> => {
      if (type === 'fake') {
        return data as FakePayload;
      }
      if (type === 'bearer') {
        try {
          return await jwtToken.verify((data as BearerPayload).hash);
        } catch (err) {
          throw YError.wrap(
            err as Error,
            'E_BAD_BEARER_TOKEN',
            type,
            (data as BearerPayload).hash,
          );
        }
      }
      throw new YError('E_UNEXPECTED_AUTH_TYPE', type);
    },
  };

  return authentication;
}
