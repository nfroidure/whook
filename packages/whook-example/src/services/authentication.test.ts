import {
  describe,
  it,
  beforeAll,
  beforeEach,
  jest,
  expect,
} from '@jest/globals';
import initAuthentication from './authentication.js';
import { YError } from 'yerror';
import initJWT from '../services/jwtToken.js';
import { type AuthenticationData } from './authentication.js';
import { type JWTService } from 'jwt-service';
import { type TimeService } from 'common-services';

describe('authentication', () => {
  const time = jest.fn<TimeService>();
  let jwtToken: JWTService<AuthenticationData>;

  beforeAll(async () => {
    jwtToken = await initJWT({
      JWT: {
        duration: '2h',
        tolerance: '15m',
        algorithms: ['HS256'],
      },
      ENV: { JWT_SECRET: 'oudelali' },
      time,
    });
  });

  beforeEach(() => {
    time.mockReset();
  });

  describe('.check()', () => {
    describe('with bearer type', () => {
      it('should work with a good token', async () => {
        time.mockReturnValueOnce(Date.parse('1982-07-22T00:00:00Z'));

        const theToken = (
          await jwtToken.sign({
            applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
            userId: 'acdc41ce-acdc-41ce-acdc-41ceacdc41ce',
            scope: 'admin',
          })
        ).token;
        const authentication = await initAuthentication({ jwtToken });

        time.mockReturnValueOnce(Date.parse('1982-07-22T01:00:00Z'));

        const result = await authentication.check('bearer', {
          hash: theToken as unknown as string,
        });

        expect({
          result,
        }).toMatchSnapshot();
      });

      it('should fail with a bad token', async () => {
        const authentication = await initAuthentication({ jwtToken });

        try {
          await authentication.check('bearer', { hash: 'lol' });
          throw new YError('E_UNEXPECTED_SUCCESS');
        } catch (err) {
          expect({
            errorCode: (err as YError).code,
            errorParams: (err as YError).params,
          }).toMatchSnapshot();
        }
      });
    });

    describe('with fake type', () => {
      it('should work with fakedata', async () => {
        const authentication = await initAuthentication({ jwtToken });
        const result = await authentication.check('fake', {
          applicationId: '1',
          userId: '1',
          scope: 'user',
        });

        expect({
          result,
        }).toMatchSnapshot();
      });
    });

    describe('with a bad auth type', () => {
      it('should fail', async () => {
        const authentication = await initAuthentication({ jwtToken });

        try {
          await authentication.check('yolo', { hash: 'lol' });
          throw new YError('E_UNEXPECTED_SUCCESS');
        } catch (err) {
          expect({
            errorCode: (err as YError).code,
            errorParams: (err as YError).params,
          }).toMatchSnapshot();
        }
      });
    });
  });
});
