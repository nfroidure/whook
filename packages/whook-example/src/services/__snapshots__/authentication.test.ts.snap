// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication .check() with a bad auth type should fail 1`] = `
{
  "errorCode": "E_UNEXPECTED_AUTH_TYPE",
  "errorParams": [
    "yolo",
  ],
}
`;

exports[`authentication .check() with bearer type should fail with a bad token 1`] = `
{
  "errorCode": "E_BAD_BEARER_TOKEN",
  "errorParams": [
    "bearer",
    "lol",
  ],
}
`;

exports[`authentication .check() with bearer type should work with a good token 1`] = `
{
  "result": {
    "applicationId": "abbacaca-abba-caca-abba-cacaabbacaca",
    "exp": 396151200,
    "iat": 396144000,
    "nbf": 396144000,
    "scope": "admin",
    "userId": "acdc41ce-acdc-41ce-acdc-41ceacdc41ce",
  },
}
`;

exports[`authentication .check() with fake type should work with fake data 1`] = `
{
  "result": {
    "applicationId": "1",
    "scope": "user",
    "userId": "1",
  },
}
`;
